### PR TITLE
Add Hardhat support

### PR DESCRIPTION
**Pull request by cgewecke**
**TODO**
- [x] Add `HARDHAT_README.md` (#567)
- [x] Investigate & resolve coverage differences between Buidler & HH (#566)
  + ConvertLib (call) is 0 vs. 100
  + EtherRouter differences  
- [x] Use higher gas limits (because some people set their defaults at MAX) (#565)
- [x] Replace death (it's not working anymore) (#565)
- [x] Add tests for more complex compiler configs (#563)
- [x] Add hardhat-e2e (#557)
  - [x] Run beta vs. [synthetix 739][1] (https://github.com/sc-forks/synthetix/pull/2)
- [x] buidler or hardhat conditional require entrypoint (for backwards compatibility) (#564)
  - [x] Rename everything (#561)
  - [x] Migrate most Buidler tests to hardhat (preserve a couple for old times) (#561)
- [x] HardhatEVM hook (#560)
- [x] HardhatPluginError (#561)
- [x] Implement/investigate #552 (configurable default account) (#562)
- [ ] Is #467 fixed by this? Check and add test....

[1]: https://github.com/Synthetixio/synthetix/pull/739

### CODE REVIEW

File: HARDHAT_README.md [hunk 1]
```diff
@@ -0,0 +1,131 @@
...
+
+```bash
+npx hardhat coverage [options]
+```
+
+| Option <img width=200/> | Example <img width=750/>| Description <img width=1000/> |
+|--------------|------------------------------------|--------------------------------|
+| testfiles  | `--testfiles "test/registry/*.ts"` | Test file(s) to run. (Globs must be enclosed by quotes.)|
+| solcoverjs | `--solcoverjs ./../.solcover.js` | Relative path from working directory to config. Useful for monorepo packages that share settings. (Path must be "./" prefixed) |
+| network    | `--network development` | Run with a ganache client over http using network settings defined in the Hardhat config. (Hardhat is the default network) |
```
Comment (cgewecke) [inline HARDHAT_README.md:49]:
```suggestion
| network    | `--network development` | Run with a ganache client over http using network settings defined in `hardhat.config.js`. (Hardhat is the default network) |
```

File: HARDHAT_README.md [hunk 2]
```diff
@@ -0,0 +1,131 @@
...
+| onTestsComplete[<sup>*</sup>][14] | *Function* |  | Hook run *after* the tests complete, *before* Istanbul reports are generated. [More...][23]|
+| onIstanbulComplete[<sup>*</sup>][14] | *Function* |  | Hook run *after* the Istanbul reports are generated, *before* the ganache server is shut down. Useful if you need to clean resources up. [More...][23]|
+
+[<sup>*</sup> Advanced use][14]
+
+## Usage
+
++ Coverage runs tests a little more slowly.
++ Coverage uses the Hardhat network by default.
++ Coverage [distorts gas consumption][13]. Tests that check exact gas consumption should be [skipped][24].
```
Comment (cgewecke) [inline HARDHAT_README.md:83]: Add note here about increasing your default account balances if you see errors like 
```
expected 'sender doesn\'t have enough funds to send tx. The upfront cost is: 72057636037927928000000000 and the sender\'s account only has: 9999999999999857947383' to include 'revert'
```

File: lib/api.js [hunk 3]
```diff
@@ -49,9 +49,8 @@ class API {
     this.skipFiles = config.skipFiles || [];
 
     this.log = config.log || console.log;
-
-    this.gasLimit = 0xffffffffff;            // default "gas sent" with transactions
-    this.gasLimitString = "0xfffffffffff";   // block gas limit for ganache (higher than "gas sent")
+    this.gasLimit = 0x1fffffffffffff;           // default "gas sent" with transactions
```
Comment (cgewecke) [inline lib/api.js:52]:
```suggestion
    this.gasLimit = 0xffffffffff;           // default "gas sent" with transactions
```

Revert this change, possibly breaking. 

Double-check that we are using gasLimitString for HH's blockGasLimit.

File: lib/api.js [hunk 4]
```diff
@@ -268,6 +267,31 @@ class API {
...
+  // Temporarily disabled because some relevant traces aren't available
+  // (maybe bytecode cannot be found)
+  hardhatTraceHandler(trace, isTraceFromCall){
+    for (const step of trace.steps){
+      if (trace.bytecode && trace.bytecode._pcToInstruction){
+        const instruction = trace.bytecode._pcToInstruction.get(step.pc)
+        this.collector.trackHardhatEVMInstruction(instruction)
+      }
+    }
+  }
```
Comment (cgewecke) [inline lib/api.js:293]: Remove hardhatTraceHandler (because currently unused)?

Comment (cgewecke) [reply lib/api.js:293]: Actually just keep - need to open repros at HH for the missing traces.

File: plugins/hardhat.plugin.js [hunk 5]
```diff
@@ -0,0 +1,213 @@
...
+    ui.report('hardhat-versions', [pkg.version]);
+
+    // Merge non-null flags into hardhatArguments
+    const flags = {};
+    for (const key of Object.keys(args)){
+      if (args[key] && args[key].length){
+        flags[key] = args[key]
+      }
+    }
+    env.hardhatArguments = Object.assign(env.hardhatArguments, flags)
```
Comment (cgewecke) [inline plugins/hardhat.plugin.js:106]: Move this whole merge-non-null to nomiclabsUtils.

File: plugins/hardhat.plugin.js [hunk 6]
```diff
@@ -0,0 +1,213 @@
...
+      tempArtifactsDir,
+      tempContractsDir
+    } = utils.getTempLocations(config);
+
+    utils.setupTempFolders(config, tempContractsDir, tempArtifactsDir)
+
+    config.paths.artifacts = tempArtifactsDir;
+    config.paths.cache = nomiclabsUtils.tempCacheDir(config);
+
+    measureCoverage = true;
```
Comment (cgewecke) [inline plugins/hardhat.plugin.js:142]: Delete line. Already set above.

File: plugins/resources/nomiclabs.ui.js [hunk 7]
```diff
@@ -65,6 +80,8 @@ class PluginUI extends UI {
     const x = ":x:";
 
     const kinds = {
+      'network-fail': `${c.red('--network argument: ')}${args[0]}` +
+                      `${c.red(' is not a defined network in hardhat.js.')}`,
```
Comment (cgewecke) [inline plugins/resources/nomiclabs.ui.js:84]:
```suggestion
                      `${c.red(' is not a defined network in hardhat.config.js.')}`,
```

File: plugins/resources/nomiclabs.utils.js [hunk 8]
```diff
@@ -0,0 +1,221 @@
...
+}
+
+function configureNetworkGas(networkConfig, api){
+  networkConfig.gas =  api.gasLimit;
+  networkConfig.gasPrice = api.gasPrice;
+}
+
+function configureHardhatEVMGas(networkConfig, api){
+  networkConfig.allowUnlimitedContractSize = true;
+  networkConfig.blockGasLimit = api.gasLimit
```
Comment (cgewecke) [inline plugins/resources/nomiclabs.utils.js:119]: Should be the `gasLimitString` string value. Is this allowed?

File: scripts/run-nomiclabs.sh [hunk 9]
```diff
@@ -76,12 +95,12 @@ echo ""
...
 cd moloch
-npm install
-npm uninstall --save-dev solidity-coverage
+npm install --silent
+npm uninstall --save-dev --silent solidity-coverage
 
 # Install and run solidity-coverage @ PR
 # Should run on network 'localhost'
-npm install --save-dev $PR_PATH
+npm install --save-dev --silent $PR_PATH
```
Comment (cgewecke) [inline scripts/run-nomiclabs.sh:103]: Doesn't silence the node-gyp noise fwiw....

File: test/integration/projects/bad-solcoverjs/hardhat.config.js [hunk 10]
```diff
@@ -0,0 +1,6 @@
+require("@nomiclabs/hardhat-truffle5");
+require(__dirname + "/../plugins/nomiclabs.plugin");
+
+module.exports={
+  logger: process.env.SILENT ? { log: () => {} } : console,
+};
```
Comment (cgewecke) [inline test/integration/projects/bad-solcoverjs/hardhat.config.js:6]: Add newline

File: test/integration/projects/ganache-solcoverjs/hardhat.config.js [hunk 11]
```diff
@@ -0,0 +1,14 @@
...
+  solidity: {
+    version: "0.5.15"
+  },
+  networks: {
+    coverage: {
+      url: "http://127.0.0.1:8555"
+    }
+  }.
+  logger: process.env.SILENT ? { log: () => {} } : console,
+};
```
Comment (cgewecke) [inline test/integration/projects/ganache-solcoverjs/hardhat.config.js:14]: ++

File: test/integration/projects/import-paths/hardhat.config.js [hunk 12]
```diff
@@ -0,0 +1,9 @@
+require("@nomiclabs/hardhat-truffle5");
+require(__dirname + "/../plugins/nomiclabs.plugin");
+
+module.exports = {
+  solidity: {
+    version: "0.5.15"
+  },
+  logger: process.env.SILENT ? { log: () => {} } : console,
+};
```
Comment (cgewecke) [inline test/integration/projects/import-paths/hardhat.config.js:9]: ++

File: test/integration/projects/libraries/hardhat.config.js [hunk 13]
```diff
@@ -0,0 +1,9 @@
+require("@nomiclabs/hardhat-truffle5");
+require(__dirname + "/../plugins/nomiclabs.plugin");
+
+module.exports = {
+  solidity: {
+    version: "0.5.15"
+  },
+  logger: process.env.SILENT ? { log: () => {} } : console,
+};
```
Comment (cgewecke) [inline test/integration/projects/libraries/hardhat.config.js:9]: ++

File: test/units/hardhat/standard.js [hunk 14]
```diff
@@ -0,0 +1,424 @@
...
+    await this.env.run("coverage");
+
+    verify.coverageGenerated(hardhatConfig);
+
+    const output = mock.getOutput(hardhatConfig);
+    const path = Object.keys(output)[0];
+    assert(
+      output[path].fnMap['1'].name === 'addTwo',
+      'cov should map "addTwo"'
+    );
```
Comment (cgewecke) [inline test/units/hardhat/standard.js:123]: Should be an expected measurement test instead of this. Could be zero.

File: test/util/integration.js [hunk 15]
```diff
@@ -32,8 +36,11 @@ function decacheConfigs(){
     `${process.cwd()}/${temp}/.solcover.js`,
     `${process.cwd()}/${temp}/${truffleConfigName}`,
     `${process.cwd()}/${temp}/${buidlerConfigName}`,
+    `${process.cwd()}/${temp}/${hardhatConfigName}`,
     `${process.cwd()}/${temp}/contracts/Simple.sol`,
-    `${process.cwd()}/${temp}/test/simple.js`
+    `${process.cwd()}/${temp}/test/simple.js`,
+    `${process.cwd()}/${temp}/test/account-one.js`,
+    `${process.cwd()}/${temp}/test/truffle-crash.js`
```
Comment (cgewecke) [inline test/util/integration.js:43]: Note to self: This is insane - adding individual files to decache.

File: plugins/resources/nomiclabs.utils.js (hunk 8 above)
Comment (cgewecke) [reply plugins/resources/nomiclabs.utils.js:119]: Not allowed
```ts
  blockGasLimit?: number;
```

File: package.json [hunk 16]
```diff
@@ -41,15 +41,19 @@
     "recursive-readdir": "^2.2.2",
     "sc-istanbul": "^0.4.5",
     "shelljs": "^0.8.3",
-    "web3": "^1.3.0"
+    "web3": "1.2.9"
```
Comment (junderw) [inline package.json:44]: Why is this set to 1.2.9?

Comment (cgewecke) [reply package.json:44]: @junderw Good catch! (And thanks for reviewing.)

Initially set this at 1.2.9 because later versions crashed Hardhat when fetching the initial accounts and client version info. Think there is a bug (or unresolved design issue) in Web3 1.3.0's implementation of the EIP1193 Provider. 

However, in a subsequent revision began making those calls with a thin wrapper in nomiclabsUtils. Web3 could be completely removed from the project if all the plugins use this...

https://github.com/sc-forks/solidity-coverage/blob/2cf65306124e594343afc80310398cf09e5b0c23/plugins/resources/nomiclabs.utils.js#L159-L165

Comment (junderw) [reply package.json:44]: ok, makes sense.

Any idea of an ETA when this will be available on npm?

Comment (cgewecke) [reply package.json:44]: Will merge and publish tomorrow (Nov 15).

**Comment by junderw** [comment]
Note: The README only refers to buidler. Perhaps it should also make mention of hardhat.

**Comment by cgewecke** [comment]
@junderw Ah yes, thanks.
